cpp/service_layer_service_impl.cc

```diff
@@ -94,3 +94,45 @@ grpc::Status ServiceLayerServiceImpl::monitor(
   }
   return grpc::Status::OK;
 }
+
+grpc::Status ServiceLayerServiceImpl::stream(
+    grpc::ServerContext* context, const chirp::StreamRequest* request,
+    grpc::ServerWriter< ::chirp::StreamReply>* stream) {
+  chirp::Timestamp initial_time;
+  std::chrono::seconds seconds =
+      std::chrono::duration_cast<std::chrono::seconds>(
+          std::chrono::system_clock::now().time_since_epoch());
+  std::chrono::microseconds useconds =
+      std::chrono::duration_cast<std::chrono::microseconds>(
+          std::chrono::system_clock::now().time_since_epoch());
+  initial_time.set_seconds(seconds.count());
+  initial_time.set_useconds(useconds.count());
+  std::set<std::string> read_chirps;
+  bool keep_streaming = true;
+  std::deque<chirp::Chirp> found_chirps;
+  while (keep_streaming) {
+    seconds = std::chrono::duration_cast<std::chrono::seconds>(
+        std::chrono::system_clock::now().time_since_epoch());
+    useconds = std::chrono::duration_cast<std::chrono::microseconds>(
+        std::chrono::system_clock::now().time_since_epoch());
+    found_chirps = service_.stream(request->hashtag(), initial_time);
+    initial_time.set_seconds(seconds.count());
+    initial_time.set_useconds(useconds.count());
+    for (chirp::Chirp c : found_chirps) {
+      if (read_chirps.find(c.id()) == read_chirps.end()) {
+        chirp::Chirp* this_chirp = new chirp::Chirp();
+        this_chirp->CopyFrom(c);
+        chirp::StreamReply reply;
+        reply.set_allocated_chirp(this_chirp);
+        chirp::StreamReply sendingReply = reply;
+        stream->Write(sendingReply);
+        read_chirps.insert(c.id());
+      }
+    }
+    if (context->IsCancelled()) {
+      keep_streaming = false;
+    }
+    usleep(20);
```

Comment:
Generally better to not have a hardcoded sleep of any kind.

Reply:
removed.